[iOS] add an unit test target to TestApp

xta0 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #29963 [iOS] update fastlane to use Scanfile
* **#29962 [iOS] add an unit test target to TestApp**

### Summary

Recently we've found that the master branch was constantly broken due to some unwanted change being landed on mobile. The problem is that our CI was not able to detect the runtime errors. Starting from this PR, we'll add some unit tests to the iOS Simulator build. As follows:

1. Add an unit test target to XCode (this PR)
2. Use Fastlane to run the tests on CI
3. Modify the CI scripts to trigger tests

### Test Plan

- Don't break the existing CI jobs unless they are flaky.

Differential Revision: [D18582908](https://our.internmc.facebook.com/intern/diff/D18582908)

xta0 committed:
[iOS] add unit test for Simulator build

shoumikhin commented on ios/TestApp/TestAppTests/TestAppTests.mm:
```diff
@@ -0,0 +1,38 @@
+
+- (void)tearDown {
+  [super tearDown];
+}
```
nit: not needed?

xta0 replied:
better to keep it, might become useful in the future when more test are added.

shoumikhin commented on ios/TestApp/TestAppTests/TestAppTests.mm:
```diff
@@ -0,0 +1,38 @@
+  auto qengines = at::globalContext().supportedQEngines();
+  if (std::find(qengines.begin(), qengines.end(), at::QEngine::QNNPACK) != qengines.end()) {
+    at::globalContext().setQEngine(at::QEngine::QNNPACK);
+  }
```
nit: looks like something that can be done once in `+setUp` instead?

shoumikhin commented on ios/TestApp/TestAppTests/TestAppTests.mm:
```diff
@@ -0,0 +1,38 @@
+  }
+  torch::autograd::AutoGradMode guard(false);
+  _module = torch::jit::load(std::string(modelPath.UTF8String));
+  _module.eval();
```
nit: maybe better to have it in each test, since some tests may actually want to change module's parameters in future?

xta0 replied:
what parameters could be changed?

shoumikhin commented on ios/TestApp/TestAppTests/TestAppTests.mm:
```diff
@@ -0,0 +1,38 @@
+#import <XCTest/XCTest.h>
+#import <torch/script.h>
```
nit: #include C++ headers?
Also, may like an LF between the two, since they belong to different types of headers

shoumikhin commented on ios/TestApp/TestAppTests/TestAppTests.mm:
```diff
@@ -0,0 +1,38 @@
+  [super setUp];
+  NSString* modelPath = [[NSBundle bundleForClass:[self class]] pathForResource:@"model"
+                                                                         ofType:@"pt"];
+  XCTAssertTrue([[NSFileManager defaultManager] fileExistsAtPath:modelPath],
```
nit: dot-syntax for the class property?

shoumikhin commented on ios/TestApp/TestAppTests/TestAppTests.mm:
```diff
@@ -0,0 +1,38 @@
+    at::globalContext().setQEngine(at::QEngine::QNNPACK);
+  }
+  torch::autograd::AutoGradMode guard(false);
+  _module = torch::jit::load(std::string(modelPath.UTF8String));
```
nit: probably no need to convert char* to std::string explicitly?

shoumikhin commented on ios/TestApp/TestAppTests/TestAppTests.mm:
```diff
@@ -0,0 +1,38 @@
+  if (std::find(qengines.begin(), qengines.end(), at::QEngine::QNNPACK) != qengines.end()) {
+    at::globalContext().setQEngine(at::QEngine::QNNPACK);
+  }
+  torch::autograd::AutoGradMode guard(false);
```
qq: wonder why this is needed?

xta0 replied:
I think this is a global switch to disable Auto grad.

shoumikhin reviewed (commented):
Did we run clang-format on it?

xta0 committed:
Update on "[iOS] add an unit test target to TestApp"


### Summary

Recently we've found that the master branch was constantly broken due to some unwanted change being landed on mobile. The problem is that our CI was not able to detect the runtime errors. Starting from this PR, we'll add some unit tests to the iOS Simulator build. As follows:

1. Add an unit test target to XCode (this PR)
2. Use Fastlane to run the tests on CI
3. Modify the CI scripts to trigger tests

### Test Plan

- Don't break the existing CI jobs unless they are flaky.

xta0 commented:
> Did we run clang-format on it?

I did.

shoumikhin reviewed (approved):
Awesome!
